src/Indicator.vala

```diff
@@ -18,51 +18,71 @@
+    public bool is_in_session { get; construct set; }
 
+    private Widgets.DisplayWidget? display_widget = null;
     private Widgets.PopoverWidget? popover_widget = null;
 
     private Services.Device primary_battery;
+    private Services.DBusInterfaces.PowerSettings screen;
+
     private bool notify_battery = false;
 
+    private int screen_brightness {
+        get { return screen.brightness; }
+        set {
+            if (screen.brightness != value) {
+                    screen.brightness = value.clamp (0, 100);
+            }
+        }
+    }
+
     public Indicator (bool is_in_session) {
         Object (code_name : Wingpanel.Indicator.POWER,
                 display_name : _("Power"),
-                description: _("Power indicator"));
-
-        this.is_in_session = is_in_session;
+                description: _("Power indicator"),
+                is_in_session: is_in_session);
     }
 
-    public override Gtk.Widget get_display_widget () {
-        if (display_widget == null) {
-            display_widget = new Widgets.DisplayWidget ();
+    construct {
+        init_bus.begin ();
+        popover_widget = new Widgets.PopoverWidget (is_in_session);
+        popover_widget.settings_shown.connect (() => close ());
+        popover_widget.brightness_change.connect ((change) => { 
+            popover_widget.slider_val += change;
+        });
+        popover_widget.brightness_new_value.connect ((new_value) => {
+            screen_brightness = new_value;
+        });
+
+        display_widget = new Widgets.DisplayWidget ();
+        display_widget.scroll_event.connect ((e) => { 
+            popover_widget.slider_val += Power.Utils.handle_scroll (e);
+        });
+
+        var dm = Services.DeviceManager.get_default ();
+
+        /* No need to display the indicator when the device is completely in AC mode */
+        if (dm.has_battery || dm.backlight.present) {
+            update_visibility ();
         }
+        dm.notify["has-battery"].connect (update_visibility);
+    }
 
+    public override Gtk.Widget get_display_widget () {
         return display_widget;
     }
 
     public override Gtk.Widget? get_widget () {
-        if (popover_widget == null) {
-            popover_widget = new Widgets.PopoverWidget (is_in_session);
-            popover_widget.settings_shown.connect (() => this.close ());
-
-            var dm = Services.DeviceManager.get_default ();
-
-            /* No need to display the indicator when the device is completely in AC mode */
-            if (dm.has_battery || dm.backlight.present) {
-                update_visibility ();
-            }
-            dm.notify["has-battery"].connect (update_visibility);
-        }
-
         return popover_widget;
     }
 
     public override void opened () {
         Services.ProcessMonitor.Monitor.get_default ().update ();
-        popover_widget.update_brightness_slider ();
+        popover_widget.slider_val = screen_brightness;
```

Comment:
when you have hotkeys on your keyboard (most laptops have)  this would result in unwanted behaviour.
because this doesn't check the updated value. it uses the local stored value.  first update the value.